File: Leetcode/Medium/4-Search-a-2D-Matrix/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows,cols = len(matrix),len(matrix[0])
        topRow,botRow = 0,rows-1
        while topRow <= botRow:
            mid = (topRow+botRow)//2
            if target > matrix[mid][-1]:
                topRow = mid + 1
            elif target < matrix[mid][0]:
                botRow = mid - 1
            else:
                break
        if not (topRow <= botRow):
            return False
        mid = (topRow+botRow)//2
        l,r = 0,cols-1
        while l<=r:
            mid2 = (l+r)//2
            if target > matrix[mid][mid2]:
                l = mid2 + 1
            elif target < matrix[mid][mid2]:
                r = mid2 - 1
            else:
                return True
        return False

File: Leetcode/Medium/4-Search-a-2D-Matrix/solution.js
let searchMatrix = function(matrix, target) {
    let [rows, cols] = [matrix.length, matrix[0].length];
    let [top, bot] = [0, rows - 1];
    while (top <= bot) {
        let mid = Math.round((top + bot) / 2);
        if (target > matrix[mid][cols - 1]) top = mid + 1;
        else if (target < matrix[mid][0]) bot = mid - 1;
        else break
    }
    if (!(top <= bot)) return false;
    let mid = Math.round((top + bot) / 2);
    let [l, r] = [0, cols - 1];
    while (l <= r) {
        let midNew = Math.round((l + r) / 2);
        if (target > matrix[mid][midNew]) l = midNew + 1;
        else if (target < matrix[mid][midNew]) r = midNew - 1;
        else return true;
    }
    return false
}